qiskit/quantum_info/synthesis/two_qubit_decompose.py

```diff
@@ -205,48 +239,376 @@ def __init__(self, unitary_matrix, eps=1e-15):
+                                       if cls is TwoQubitWeylDecomposition else cls)
+        elif is_close(0, 0, 0):
+            instance = super().__new__(TwoQubitWeylIdEquiv)
+        elif is_close(pi4, pi4, pi4) or is_close(pi4, pi4, -pi4):
+            instance = super().__new__(TwoQubitWeylSWAPEquiv)
+        elif (lambda x: is_close(x, x, x))(_closest_partial_swap(a, b, c)):
+            instance = super().__new__(TwoQubitWeylPartialSWAPEquiv)
+        elif (lambda x: is_close(x, x, -x))(_closest_partial_swap(a, b, -c)):
+            instance = super().__new__(TwoQubitWeylPartialSWAPFlipEquiv)
+        elif is_close(a, 0, 0):
+            instance = super().__new__(TwoQubitWeylControlledEquiv)
+        elif is_close(pi4, pi4, c):
+            instance = super().__new__(TwoQubitWeylMirrorControlledEquiv)
+        elif is_close((a+b)/2, (a+b)/2, c):
+            instance = super().__new__(TwoQubitWeylfSimaabEquiv)
+        elif is_close(a, (b+c)/2, (b+c)/2):
+            instance = super().__new__(TwoQubitWeylfSimabbEquiv)
+        elif is_close(a, (b-c)/2, (c-b)/2):
+            instance = super().__new__(TwoQubitWeylfSimabmbEquiv)
+        else:
+            instance = super().__new__(TwoQubitWeylGeneral)
+
+        instance._original_decomposition = od
+        return instance
+
+    def __init__(self, unitary_matrix, fidelity=None):
+        del unitary_matrix  # unused in __init__ (used in new)
+        od = self._original_decomposition
+        self.a, self.b, self.c = od.a, od.b, od.c
+        self.K1l, self.K1r = od.K1l, od.K1r
+        self.K2l, self.K2r = od.K2l, od.K2r
+        self.global_phase = od.global_phase
+        self.unitary_matrix = od.unitary_matrix
+        self.requested_fidelity = fidelity
+        self._is_flipped_from_original = False
+        self.specialize()
+
+        # Update the phase after specialization:
+        if self._is_flipped_from_original:
+            da, db, dc = (np.pi/2-od.a)-self.a, od.b-self.b, -od.c-self.c
+            tr = 4 * complex(math.cos(da)*math.cos(db)*math.cos(dc),
+                             math.sin(da)*math.sin(db)*math.sin(dc))
+        else:
+            da, db, dc = od.a-self.a, od.b-self.b, od.c-self.c
+            tr = 4 * complex(math.cos(da)*math.cos(db)*math.cos(dc),
+                             math.sin(da)*math.sin(db)*math.sin(dc))
+        self.global_phase += cmath.phase(tr)
+        self.calculated_fidelity = trace_to_fid(tr)
+        if logger.isEnabledFor(logging.DEBUG):
+            actual_fidelity = self.actual_fidelity()
+            logger.debug("Requested fidelity: %s calculated fidelity: %s actual fidelity %s",
+                         self.requested_fidelity, self.calculated_fidelity, actual_fidelity)
+            if abs(self.calculated_fidelity - actual_fidelity) > 1.E-12:
+                logger.warning("Requested fidelity different from actual by %s",
+                               self.calculated_fidelity - actual_fidelity)
```

Comment:
Does this actually get exercised in the tests?  By default I think the test env doesn't set a log level. Do you think it makes sense to set debug log level for the test classes in test_synthesis to ensure this gets run and we can check for it in the tests?

Reply:
good idea i'll add that.
I was looking at coverage and i'll also add a test for __repr__ just for completeness,